conduit/pipeline/pipeline.go

```diff
@@ -431,103 +445,243 @@ func addMetrics(block data.BlockData, importTime time.Duration) {
 	metrics.ImportedTxnsPerBlock.Observe(float64(len(block.Payset)) + float64(innerTxn))
 }
 
-// Start pushes block data through the pipeline
-func (p *pipelineImpl) Start() {
+func (p *pipelineImpl) importerHandler(importer importers.Importer, roundChan <-chan uint64, blkOutChan pluginChannel) {
 	p.wg.Add(1)
-	retry := uint64(0)
 	go func() {
 		defer p.wg.Done()
-		// We need to add a separate recover function here since it launches its own go-routine
-		defer HandlePanic(p.logger)
+		lastRnd := uint64(0)
+		totalSelectWait := time.Duration(0)
+		totalFeedWait := time.Duration(0)
+		defer func() {
+			p.logger.Debugf("importer handler exiting. lastRnd=%d totalSelectWait=%dms; totalFeedWait=%dms", lastRnd, totalSelectWait.Milliseconds(), totalFeedWait.Milliseconds())
+		}()
+
 		for {
-		pipelineRun:
-			metrics.PipelineRetryCount.Observe(float64(retry))
-			if retry > p.cfg.RetryCount && p.cfg.RetryCount != 0 {
-				p.logger.Errorf("Pipeline has exceeded maximum retry count (%d) - stopping...", p.cfg.RetryCount)
+			startRound := time.Now()
+			select {
+			case <-p.ctx.Done():
+				p.logger.Infof("importer handler exiting. lastRnd=%d", lastRnd)
 				return
+			case rnd := <-roundChan:
+				lastRnd = rnd
+				waitTime := time.Since(startRound)
+				totalSelectWait += waitTime
+				p.logger.Tracef("importer handler waited %dms to receive round %d", waitTime.Milliseconds(), rnd)
+
+				blkData, importTime, lastError := Retries(importer.GetBlock, rnd, p, importer.Metadata().Name)
+				if lastError != nil {
+					p.cancelWithProblem(fmt.Errorf("importer %s handler (%w): failed to import round %d after %dms: %w", importer.Metadata().Name, errImporterCause, rnd, importTime.Milliseconds(), lastError))
+					return
+				}
+				metrics.ImporterTimeSeconds.Observe(importTime.Seconds())
+				p.logger.Tracef("importer handler imported round %d in %dms", rnd, importTime.Milliseconds())
+
+				// TODO: Verify that the block was built with a known protocol version.
+
+				importFinish := time.Now()
+
+				select {
+				case <-p.ctx.Done():
+					return
+				case blkOutChan <- blkData:
+					waitTime := time.Since(importFinish)
+					totalFeedWait += waitTime
+					p.logger.Tracef("imported round %d into blkOutChan after waiting %dms on channel", rnd, waitTime.Milliseconds())
+				}
```

Comment:
Also in the other handlers, this could be helpful
```suggestion
				// check context in case out channel is full.
				select {
				case <-p.ctx.Done():
					return
				case blkOutChan <- blkData:
					waitTime := time.Since(importFinish)
					totalFeedWait += waitTime
					p.logger.Tracef("imported round %d into blkOutChan after waiting %dms on channel", rnd, waitTime.Milliseconds())
				}
```